### PR TITLE
Added the sources to review

### PR DESCRIPTION
**Pull request by holyjak**

### CODE REVIEW

File: main/java/us/klette/constantstring/CString.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+ * set of operations.
+ *
+ * @author Kristian Klette (klette@klette.us)
+ */
+public interface CString {
+
+    /**
+     * Creates a new CString instance from a String.
+     *
+     * @param value The String to use as a base value
```
Comment (holyjak) [inline main/java/us/klette/constantstring/CString.java:50]: I think the @param, @return docs add no value and would remove them (until you're just making Checkstyle happy - in which case I'd anyway prefer to fix Checkstyle config) I'll not comment on this topic at other places.

File: main/java/us/klette/constantstring/CString.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+     * @return The resulting {@link String} instance.
+     */
+    @Override
+    String toString();
+
+    /**
+     * Applies a substring operations similar to
+     * {@link java.lang.String#substring(int)}.
+     *
+     * @param index The index of which to start the new string
```
Comment (holyjak) [inline main/java/us/klette/constantstring/CString.java:69]: is `index` inclusive or exclusive? why not just copy javadoc from String.substring for this?

> beginIndex - the beginning index, inclusive.

File: main/java/us/klette/constantstring/CString.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    /**
+     * Applies a substring operations similar to
+     * {@link java.lang.String#substring(int, int)}.
+     *
+     * @param index The index of which to start the new string
+     * @param end The end index of the substring
+     * @return The CString representation after the applied substring operation
+     */
+    default CString substring(final int index, final int end) {
+        return new CStringSubRange(this, index, end);
```
Comment (holyjak) [inline main/java/us/klette/constantstring/CString.java:84]: Not sure if we need both `CStringSub` and `CStringSubRange,` isn't it the same case, `substring(3)` being simply `substring(3, this.length)`? (plus minus 1 :))

File: main/java/us/klette/constantstring/CString.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+     * Delete a subsection of the String.
+     *
+     * @param start The start index of the delete
+     * @param end The end index of the deleted part
+     * @return The CString representation with the delete applied.
+     */
+    default CString delete(final int start, final int end) {
+        return new CStringConcat(
+                new CStringSubRange(this, 0, start),
+                new CStringSub(this, start+end)
```
Comment (holyjak) [inline main/java/us/klette/constantstring/CString.java:116]: hm, why isn't it just `concat(substring(0, start), substring(end))`? (not sure about inclusive/enclusive and thus +-1)
I.e. 
1. why the Sub is for `start+end`?!  If I have "aaaSbbbEccc" and want "aaaccc"? Perhaps `end` is actualy `length`?!
2. using the existing methods would make it easier for me to understand, though it is a secondary point

File: main/java/us/klette/constantstring/CStringLeaf.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+ * Represents a value node of a CString operation tree.
+ *
+ * @author Kristian Klette (klette@klette.us)
+ */
+final class CStringLeaf implements CString {
+
+    /**
+     * The value this leaf nodes carries.
+     */
+    private final transient String value;
```
Comment (holyjak) [inline main/java/us/klette/constantstring/CStringLeaf.java:44]: `transient` is so rarely used that you should really explain the need for it, especially on a `final` var

File: main/java/us/klette/constantstring/CStringSub.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    /**
+     * Creates a new substring operation on the given value.
+     *
+     * @param val   The value used as the child node for the operation.
+     * @param index The index from which the text is kept.
+     */
+    public CStringSub(@Nonnull final CString val, final int index) {
+        this.value = val;
+        this.idx = index;
```
Comment (holyjak) [inline main/java/us/klette/constantstring/CStringSub.java:59]: Mixing `idx` and `index` = using 2 names for 1 concept, which is bad. Don't be lase, call it `beginIndex` as JDK does :)

File: main/java/us/klette/constantstring/CStringSub.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    @Override
+    public final String toString() {
+        final String childEval = this.value.toString();
```
Comment (holyjak) [inline main/java/us/klette/constantstring/CStringSub.java:64]: I dislike the `childEval` name but can't think of a perfect one either. `valueAsString` or `valueString` would be somewhat better (we at least keep the name of the concept here, value, and add the info how it differs from this.value). "Eval" feels somewhat unclear and does not speak about what the value is but what was done to it.

File: main/java/us/klette/constantstring/CStringSubRange.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package us.klette.constantstring;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Representation of a substring operation.
```
Comment (holyjak) [inline main/java/us/klette/constantstring/CStringSubRange.java:36]: Now you should really explain how it relates to and differs from `CStringSub`

File: main/java/us/klette/constantstring/CStringSubRange.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+     */
+    private final transient CString value;
+
+    /**
+     * The index from which the text is kept.
+     */
+    private final transient int idx;
+
+    /**
+     * The number of chars after {@link #idx} to include.
```
Comment (holyjak) [inline main/java/us/klette/constantstring/CStringSubRange.java:53]: Nooo! How can you call it `end`?! In JDK end, or rather endIndex is actually an index. If you want "number of chars" then you - well, at least JDK - call it `length`. #consistency #principleOfLastSurprise

File: main/java/us/klette/constantstring/package-info.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+ * THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package us.klette.constantstring;
```
Comment (holyjak) [inline main/java/us/klette/constantstring/package-info.java:30]: Empty files? WTF?

File: test/java/us/klette/constantstring/CStringLeafTest.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package us.klette.constantstring;
+
+import org.junit.Test;
+
+public class CStringLeafTest {
+    @Test(expected = NullPointerException.class)
+    public void shouldThrowNPEIfConstructedWithNull() throws Exception {
```
Comment (holyjak) [inline test/java/us/klette/constantstring/CStringLeafTest.java:37]: Maybe you record too much impl. details in the name? If you switch to a different exception (f.e.x. IllegalArgument which I prefer) you;d need to rename it. I'd prefer the more business-logic focused

```
cannotConstructCStringFromNull
```

or st. along the line

File: test/java/us/klette/constantstring/CStringLeafTest.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+        new CStringLeaf(null);
+    }
+}
```
Comment (holyjak) [inline test/java/us/klette/constantstring/CStringLeafTest.java:40]: Maybe we could also prove we can actually construct some non-null instance? And a question that immediatelly pops out, what about ""?

File: test/java/us/klette/constantstring/CStringSubRangeTest.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class CStringSubRangeTest {
+
+    @Test
+    public void testToString() throws Exception {
+        CStringSubRange foobar = new CStringSubRange(CString.create("foobar"), 3, 4);
+        String s = foobar.toString();
+        assertThat(s).isEqualTo("foobar".substring(3, 4));
+    }
```
Comment (holyjak) [inline test/java/us/klette/constantstring/CStringSubRangeTest.java:44]: I really dislike this one:
- Name - no reference to actual business requirement
- "foobar" is name with 0 domain knowledge value, as is "s"
- I do not understand why you picked the indices 3 and 4, what is their relation to the value and expected result
- Also, wouldn't it be better to use the public API of the lib, i.e. create a CString and call substring on it? Thus you're free to rename your classes and lot other without the test becoming outdated; the abstraction level of the test becomes higher, closer to business
  (I wanted to propose a better name but I really fail to understand why substring(.., 3,4) should return the string unchanged)

File: test/java/us/klette/constantstring/CStringSubRangeTest.java [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+    @Test
+    public void overflow() throws Exception {
```
Comment (holyjak) [inline test/java/us/klette/constantstring/CStringSubRangeTest.java:47]: The name provides way to little context, too far from biz requirement lang. =>

```
shouldReturnEmptyStringIfIndexAtOrBeyondEnd
```

Also I suppose we should somewhere explain why it is not a failure to go beyond the end

File: test/java/us/klette/constantstring/CStringSubRangeTest.java [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+        CStringSubRange foobar = new CStringSubRange(CString.create("foobar"), 10, 12);
+        String s = foobar.toString();
+        assertThat(s).isEqualTo("");
+    }
+
+    @Test
+    public void partialOverflow() throws Exception {
```
Comment (holyjak) [inline test/java/us/klette/constantstring/CStringSubRangeTest.java:54]: same as above =>

```
shouldReturnSubstringFromIndexToEndIfRequestedMoreThanAvailable
```

File: test/java/us/klette/constantstring/OperationsTest.java [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package us.klette.constantstring;
+
+import org.junit.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Test cases for basic operations.
```
Comment (holyjak) [inline test/java/us/klette/constantstring/OperationsTest.java:37]: It's 'OperationsTest' but you call it "basic operations" so either drop "basic" or call it `BasicOperationsTest` #consistency

File: test/java/us/klette/constantstring/OperationsTest.java [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+ *
+ * @author Kristian Klette (klette@klette.us)
+ */
+public final class OperationsTest {
+
+    @Test
+    public void create() throws Exception {
```
Comment (holyjak) [inline test/java/us/klette/constantstring/OperationsTest.java:44]: `shouldReturnSameStringAsCreatedFrom`?

File: test/java/us/klette/constantstring/OperationsTest.java [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+        CString foobar = CString.create("foobar");
+
+        assertThat(foobar.toString()).isEqualTo("foobar");
+    }
+
+    /**
+     * Verify that CString can store and evaluate basic strings.
+     *
+     * @throws Exception e
```
Comment (holyjak) [inline test/java/us/klette/constantstring/OperationsTest.java:53]: thank you for this useful javadoc line #irony ;)

File: test/java/us/klette/constantstring/OperationsTest.java [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+     */
+    @Test
+    public void evalShouldReturnBasicValue() throws Exception {
```
Comment (holyjak) [inline test/java/us/klette/constantstring/OperationsTest.java:56]: Isn't this the same as the previous one? Plus
- WTF is "eval", there is no such method?!
- WTF is "basic value"?!

File: test/java/us/klette/constantstring/OperationsTest.java [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+     * @throws Exception e
+     */
+    @Test
+    public void shouldBeAbleToSubstring() throws Exception {
+        final String string = "substr";
+        final int index = 3;
+        final CString initial = CString.create(string);
+        final CString substring = initial.substring(index);
+        final String evaluated = substring.toString();
+        assertThat(evaluated).isEqualTo(string.substring(index));
```
Comment (holyjak) [inline test/java/us/klette/constantstring/OperationsTest.java:75]: It looks like you are paied by LOC :-)

``` java
final int index = 3;
final String original = "some text";
assertThat(new CString(original).substring(index).toString()).equals(original.substring(index));
```

And what about `substringWithoutLengthProducesSameResultAsJavaSubstring`?

File: test/java/us/klette/constantstring/OperationsTest.java [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+    /**
+     * Verify that substrings of substrings are working.
+     *
+     * @throws Exception e
+     */
+    @Test
+    public void substringOfSubstring() throws Exception {
+        final CString initial = CString.create("abc");
+        final CString first = initial.substring(1);
+        final CString second = first.substring(1);
```
Comment (holyjak) [inline test/java/us/klette/constantstring/OperationsTest.java:102]: I have to map mentally what 1st/2nd substring corresponds to. I might prefer `substringBC` and `substringC`

File: test/java/us/klette/constantstring/OperationsTest.java [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+        final String evaluated = second.toString();
+        assertThat(evaluated).isEqualTo("c");
+    }
+
+    @Test
+    public void concatTest() throws Exception {
```
Comment (holyjak) [inline test/java/us/klette/constantstring/OperationsTest.java:108]: shouldConcatTwoCStrings ?

File: test/java/us/klette/constantstring/OperationsTest.java [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+        final CString foo = CString.create("foo");
+        final CString bar = CString.create("bar");
+        final CString foobar = foo.concat(bar);
+
+        assertThat(foobar.toString()).isEqualTo("foobar");
+    }
+    @Test
+    public void concatStringTest() throws Exception {
```
Comment (holyjak) [inline test/java/us/klette/constantstring/OperationsTest.java:116]: shouldConcatStringOntoCString ?

File: test/java/us/klette/constantstring/OperationsTest.java [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+        final CString foo = CString.create("foo");
+        final CString foobar = foo.concat("bar");
+
+        assertThat(foobar.toString()).isEqualTo("foobar");
+    }
+
+    @Test
+    public void delete() throws Exception {
```
Comment (holyjak) [inline test/java/us/klette/constantstring/OperationsTest.java:124]: I miss deleting also in the middle, at the end and perhaps what if I try to delete beyond the end. Also, you sometimes use "should" sometimes not in the test names - WTF consistency?

```
shouldDeleteLengthCharsFromIndex
```

File: test/java/us/klette/constantstring/OperationsTest.java [hunk 25]
```diff
@@ -0,0 +1,156 @@
...
+        CString foobar = CString.create("foobar");
+        CString bar = foobar.delete(0, 3);
+
+        assertThat(bar.toString()).isEqualTo("bar");
+    }
+
+    @Test
+    public void substring() throws Exception {
```
Comment (holyjak) [inline test/java/us/klette/constantstring/OperationsTest.java:132]: haven't I seen this test before?? And aren't there multiple dufferent cases of substrings? Do we test all of them here?

File: test/java/us/klette/constantstring/SubstringTest.java [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+     *
+     * @return The collection of assert objects.
+     */
+    @Parameterized.Parameters(
+            name = "{index}: {0}.substring({1}) == \"{2}\"")
+    public static Collection<Object[]> data() {
+        final String base = "abc";
+        return Arrays.asList(
+                new Object[][]{
+                        {base, 0, base},
```
Comment (holyjak) [inline test/java/us/klette/constantstring/SubstringTest.java:84]: sometimes you use "base" sometimes "initial" #consistency

And would it be so much work to hardcode the expected values as you do for idx=4? Would be perhaps more readable

File: test/java/us/klette/constantstring/SubstringTest.java [hunk 27]
```diff
@@ -0,0 +1,105 @@
...
+        );
+    }
+
+    /**
+     * Run the assert against all paramters.
+     *
+     * @throws Exception e
+     */
+    @Test
+    public final void substringTest() throws Exception {
```
Comment (holyjak) [inline test/java/us/klette/constantstring/SubstringTest.java:99]:
```
[substring]shouldReturnFromIndexToEnd
```

File: main/java/us/klette/constantstring/CString.java (hunk 1 above)
Comment (klette) [inline main/java/us/klette/constantstring/CString.java:50]: Except for some nicer styling in your IDE of choice. But I see your point

File: main/java/us/klette/constantstring/CStringSub.java (hunk 7 above)
Comment (klette) [inline main/java/us/klette/constantstring/CStringSub.java:64]: I was really in doubt about this one. The point is to mark this operation as "dangerous" as its linear time, since the tree is evaluated/computed/traversed.

It might be a bit unclear, but each CString-instance is a node in the computational graph of the final result.

Comment (holyjak) [inline main/java/us/klette/constantstring/CStringSub.java:64]: I see your point though I don't think that using the name to indicate it is a good solution. It feels like leveraging an existing bug to cause a problem caused by another bug instead of fixing the original bug in the first place :-)
It might be better to make toString final in the CString interface (if possible?) and add `protected String evaluate()` and use that everywhere internally. The single toString would just call it. Then you don't need to misues a variable to draw attention to the fact the method is doing something and is expensive.
